docs/proposals/dynamic-app-parameters.md

```diff
@@ -0,0 +1,181 @@
+
+In a typical Argo CD installation however, the component that’s used to render Helm charts (the repository server) does not have access to nor contextual knowledge (beside a few informational assets such cluster version and APIs provided) about the cluster where the chart is going to be installed to. Argo CD performs an offline templating of the chart, so there is no way for this kind of dynamic parameterization to work.
+
+The current workaround is to parameterize all Argo CD Applications using explicit, application specific parameters. This can become tedious and cumbersome, especially when using mechanisms such as ApplicationSet to render large amounts of Applications. Also, secret data is not to be stored in Application manifests, since these objects are considered non-confidential.
+
+The community wants to leverage the functionality behind `helm lookup` with Argo CD. However natively supporting `helm lookup` would require a significant architectural change, and it would also have severe security implications to be resolved.
+
+## Non-goals
+
+* Supporting `helm lookup` “as is” is a non-goal. Example, `helm lookup` constructs like `“lookup "v1" "Pod" "mynamespace" "mypod"` are not supported. This proposal addresses an Argo CD-native mechanism to support consumption of in-cluster information in helm charts and possibly other types of Applications in the future. 
+* Emulate or otherwise make this feature a replacement for secrets management (albeit it may allow Applications to read values stored in secrets on the cluster)
+
+## Proposal
+
+The authors of this document propose the introduction of a new kind of parameter for Helm applications (and later possibly other types of Applications): The dynamic parameter.
+
+Currently, Helm Applications within Argo CD support _statically defined_ parameters, which are simple key/value mappings and which need to be set to the desired value before the Application's manifests are being rendered.
+
+A dynamic parameter is defined similarly to how a static parameter is defined currently, but instead of supplying a static value, it is configured using a reference to a resource that exists in the destination cluster. 
+
+For example:
+
+```yaml
+kind: Application
+apiVersion: argoproj.io/v1alpha1
+metadata:
+  name: foo-app
+spec:
+  project: default
+  source:
+    repoURL: https://some.helm.repo.com
+    chart: foo
+    targetRevision: 1.0.x
+    helm:
+      parameters:
+      - name: foo
+        value: bar
+      dynamicParameters:
```

Comment:
```suggestion
      valuesFrom:
```
I proposed a different name for better UX

Reply:
Hm. Are you suggesting to replace  `dynamicParameters` by `valuesFrom`, or to use `valueFrom` within parameters, i.e. so that one could have:

```
parameters
- name: param1
  value: some-static-value
- name: parame2
  valueFrom:
    resourceRef:
    ...
```

Reply:
i think the `valuesFrom` is more `kubernetes style` of reference doing it even using the `volumesFrom` will be a intuitive way of using the feature

Reply:
@jannfis I was thinking in terms of the string/name only `s/dynamicParameters/valuesFrom`
But also I'm fine with  `dynamicParameters` or `dynamicValues`

What I feel strong is about @crenshaw-dev [comment](https://github.com/argoproj/argo-cd/pull/12050#issuecomment-1564909533) on making the feature work not just for helm but for kustomize and plugin also, the a better name would be `envFrom`

```
spec:
  source:
    envFrom:
      - name: bar
        resourceRef:
          kind: ConfigMap
          name: some-cm 
          group: "" # if omitted, use ""
          namespace: some-ns # if omitted, use target namespace
          path: .data.somekey # if omitted, set param to true if resource exists or to false if resource does not exist
    helm:
      parameters:
        - name: bar
          value: "$ENV_BAR"
    kustomize:
      images:
      - image=image:$ENV_BAR
    plugin:
      env:
      - name: BAR
        value: "$ENV_BAR"
```

Reply:
Yeah I need to write up my thoughts after 2.8rc1 is out. Jann's points against using env var interpolation are good. I still slightly favor that way, but tbh anything that is usable for Kustomize and plugins are fine with me, even if it's not built into v1. As long as the code is reasonably modular and reusable.

Reply:
The 2.9 cycle will mostly be tech debt and vacation for me. If Jann wants to push forward with the proposal in the meantime, I'll trust his instinct on the design. :-)

Reply:
Being a bit late to the chat, just would like to point out that `valuesFrom:` is exactly what flux-cd uses for similar behaviour.
Been trying to support the requests for something like this (https://github.com/argoproj/argo-cd/issues/1786#issuecomment-2365132865, https://github.com/argoproj/argo-cd/issues/1786#issuecomment-2365301701)